ext/src/autobahn/encoder/MsgPackLz4Encoder.java

```diff
@@ -13,61 +14,43 @@
 import com.headius.jruby.lz4.vendor.net.jpountz.lz4.LZ4Factory;
 
+import org.msgpack.jruby.Encoder;
+import org.msgpack.jruby.Decoder;
+
 
 @JRubyClass(name="Autobahn::MsgPackLz4Encoder")
-public class MsgPackLz4Encoder extends RubyObject {
+public class MsgPackLz4Encoder extends MsgPackEncoderBase {
   private static final LZ4Factory LZ4_FACTORY = LZ4Factory.fastestInstance();
-  private static final RubyString CONTENT_TYPE = Ruby.getGlobalRuntime().newString("application/msgpack");
   private static final RubyString CONTENT_ENCODING = Ruby.getGlobalRuntime().newString("lz4");
 
   private final LZ4Compressor compressor;
   private final LZ4Decompressor decompressor;
-  private final MessagePack msgPack;
-  private final RubyObjectPacker packer;
-  private final RubyObjectUnpacker unpacker;
-  private final RubyHash properties;
-  private RubyHash unpackerOptions;
 
   public MsgPackLz4Encoder(Ruby runtime, RubyClass type) {
-    super(runtime, type);
+    super(runtime, type, CONTENT_ENCODING);
     this.compressor = LZ4_FACTORY.fastCompressor();
     this.decompressor = LZ4_FACTORY.decompressor();
-    this.msgPack = new MessagePack();
-    this.packer = new RubyObjectPacker(msgPack);
-    this.unpacker = new RubyObjectUnpacker(msgPack);
-    this.properties = RubyHash.newHash(runtime);
-    this.properties.put(runtime.newSymbol("content_type"), CONTENT_TYPE);
-    this.properties.put(runtime.newSymbol("content_encoding"), CONTENT_ENCODING);
   }
 
-  @JRubyMethod(name = "initialize", optional = 1, visibility = PRIVATE)
-  public IRubyObject initialize(ThreadContext ctx, IRubyObject[] args) {
-    unpackerOptions = (args.length == 1 && args[0] instanceof RubyHash) ? (RubyHash) args[0] : null;
-    return this;
+  @JRubyMethod(name = "content_encoding", module = true)
+  public static IRubyObject getContentEncoding(ThreadContext ctx, IRubyObject recv) {
+    return CONTENT_ENCODING;
   }
 
-  @JRubyMethod(required = 1)
-  public IRubyObject encode(ThreadContext ctx, IRubyObject obj) throws IOException {
-    byte[] packed = packer.packRaw(obj);
-    int maxBufferSize = 4 + compressor.maxCompressedLength(packed.length);
+  protected RubyString compress(Ruby runtime, ByteList packed) throws IOException {
+    int maxBufferSize = 4 + compressor.maxCompressedLength(packed.length());
     byte[] compressed = new byte[maxBufferSize];
-    int headerSize = encodeHeader(packed.length, compressed);
-    int compressedSize = compressor.compress(packed, 0, packed.length, compressed, headerSize, maxBufferSize - headerSize);
-    return RubyString.newStringNoCopy(ctx.getRuntime(), compressed, 0, headerSize + compressedSize);
+    int headerSize = encodeHeader(packed.length(), compressed);
+    int compressedSize = compressor.compress(packed.unsafeBytes(), packed.begin(), packed.length(), compressed, headerSize, maxBufferSize - headerSize);
+    return RubyString.newStringNoCopy(runtime, compressed, 0, headerSize + compressedSize);
```

Comment:
I guess the assumption is that the compressed buffer will not stick around for too long. If it would, it would potentially hold on to a much too large buffer. One could consider allocating a new string with a copy in the case that compressedSize is much smaller than maxBufferSize.